new-tutorial/L4_message_passing.py

```diff
@@ -0,0 +1,386 @@
+"""
+Writing GNN Modules for Stochastic GNN Training
+===============================================
+
+All GNN modules DGL provides support stochastic GNN training. This
+tutorial teaches you how to write your own graph neural network module
+for stochastic GNN training. It assumes that
+
+1. You know :doc:`how to write GNN modules for full graph
+   training <3_message_passing>`.
+2. You know :doc:`how stochastic GNN training pipeline
+   works <L1_large_node_classification>`.
+
+"""
+
+import dgl
+import torch
+import numpy as np
+from ogb.nodeproppred import DglNodePropPredDataset
+
+dataset = DglNodePropPredDataset('ogbn-products')
+
+graph, node_labels = dataset[0]
+idx_split = dataset.get_idx_split()
+train_nids = idx_split['train']
+node_features = graph.ndata['feat']
+
+sampler = dgl.dataloading.MultiLayerNeighborSampler([4, 4])
+train_dataloader = dgl.dataloading.NodeDataLoader(
+    graph, train_nids, sampler,
+    batch_size=1024,
+    shuffle=True,
+    drop_last=False,
+    num_workers=0
+)
+
+input_nodes, output_nodes, bipartites = next(iter(train_dataloader))
+
+
+######################################################################
+# DGL Bipartite Graph Introduction
+# --------------------------------
+# 
+# In the previous tutorials, you have seen the concept *bipartite graph*.
+# This section introduces how you can manipulate (directional) bipartite
+# graphs.
+# 
+# You can access the input node features and output node features via
+# ``srcdata`` and ``dstdata`` attributes:
+# 
+
+bipartite = bipartites[0]
+print(bipartite.srcdata)
+print(bipartite.dstdata)
+
+
+######################################################################
+# It also has ``num_src_nodes`` and ``num_dst_nodes`` functions to query
+# how many input nodes and output nodes exist in the bipartite graph:
```

Comment:
input/output -> source/destination

Reply:
Not changing it for now with the same rationale as above.